components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/token/OAuth2TokenEndpoint.java

```diff
@@ -67,202 +71,188 @@
 
-            if (request.getHeader(OAuthConstants.HTTP_REQ_HEADER_AUTHZ) != null) {
-
-                try {
-                    String[] clientCredentials = EndpointUtil.extractCredentialsFromAuthzHeader(
-                            request.getHeader(OAuthConstants.HTTP_REQ_HEADER_AUTHZ));
+            if (oauth2AccessTokenResp.getErrorMsg() != null) {
+                return handleErrorResponse(oauth2AccessTokenResp);
+            } else {
+                return buildTokenResponse(oauth2AccessTokenResp);
+            }
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
 
-                    // The client MUST NOT use more than one authentication method in each request
-                    if (paramMap.containsKey(OAuth.OAUTH_CLIENT_ID)
-                            && paramMap.containsKey(OAuth.OAUTH_CLIENT_SECRET)) {
-                        return handleBasicAuthFailure();
-                    }
+    }
 
-                    //If a client sends an invalid base64 encoded clientid:clientsecret value, it results in this
-                    //array to only contain 1 element. This happens on specific errors though.
-                    if (clientCredentials.length != 2) {
-                        return handleBasicAuthFailure();
-                    }
+    private CarbonOAuthTokenRequest buildCarbonOAuthTokenRequest(HttpServletRequestWrapper httpRequest)
+            throws OAuthSystemException, TokenEndpointBadRequestException {
 
-                    // add the credentials available in Authorization header to the parameter map
-                    paramMap.add(OAuth.OAUTH_CLIENT_ID, clientCredentials[0]);
-                    paramMap.add(OAuth.OAUTH_CLIENT_SECRET, clientCredentials[1]);
+        try {
+            return new CarbonOAuthTokenRequest(httpRequest);
+        } catch (OAuthProblemException e) {
+            return handleInvalidRequest(e);
+        }
+    }
 
-                    consumerKey = clientCredentials[0];
+    private CarbonOAuthTokenRequest handleInvalidRequest(OAuthProblemException e) throws TokenEndpointBadRequestException {
 
-                } catch (OAuthClientException e) {
-                    // malformed credential string is considered as an auth failure.
-                    log.error("Error while extracting credentials from authorization header", e);
-                    return handleBasicAuthFailure();
-                }
-            } else if (StringUtils.isNotEmpty(httpRequest.getParameter(OAuth.OAUTH_CLIENT_ID))) {
-                consumerKey = httpRequest.getParameter(OAuth.OAUTH_CLIENT_ID);
+        /*Since oltu library sends OAthProblemException upon real exception and input errors we need to show
+            input errors when debugging and need to show the error logs when real exception thrown
+        */
+        if (OAuthError.TokenResponse.INVALID_REQUEST.equalsIgnoreCase(e.getError()) ||
```

Comment:
Can we segregate these two if conditions and log accordingly. This will help on analyzing the logs.